include/circt/Support/CPPCDE.h

```diff
@@ -0,0 +1,398 @@
+                          ESICDEValue(this->ctx, unwrapOp.getResult(1)));
+  }
+};
+
+struct DefaultCDEValue
+    : public CDEValueImpl<CDEOperatorSelector, DefaultCDEValue> {
+  using CDEValueImpl::CDEValueImpl;
+};
+
+// A class for providing backedge-based access to the in- and output ports of
+// a module.
+template <typename CDEValue>
+class CDEPorts {
+
+public:
+  CDEPorts(CDEContext &ctx, hw::HWModuleOp module) : ctx(ctx) {
+    auto modPorts = module.getPorts();
+
+    for (auto [barg, info] : llvm::zip(module.getArguments(), modPorts.inputs))
+      ports.try_emplace(info.name.str(), CDEValue(&ctx, barg));
+
+    auto outputOp = cast<hw::OutputOp>(module.getBodyBlock()->getTerminator());
+    assert(outputOp.getNumOperands() == 0);
+
+    OpBuilder::InsertionGuard g(ctx.b);
+    ctx.b.setInsertionPoint(outputOp);
+    llvm::SmallVector<Value> outputOpArgs;
+    for (auto &info : modPorts.outputs) {
+      outputBackedges.push_back(ctx.bb.get(info.type));
+      ports.try_emplace(info.name.str(),
+                        CDEValue(&ctx, &outputBackedges.back()));
+      outputOpArgs.push_back(outputBackedges.back());
+    }
+
+    ctx.b.create<hw::OutputOp>(outputOp.getLoc(), outputOpArgs);
+    outputOp.erase();
+  }
+
+  CDEValue operator[](llvm::StringRef name) {
+    auto it = ports.find(name.str());
+    assert(it != ports.end() && "Port not found.");
+    return it->second;
+  }
+
+private:
+  std::map<std::string, CDEValue> ports;
+  CDEContext &ctx;
+  std::vector<Backedge> outputBackedges;
+};
+
+// A CDE Generator is a utility class for supporting syntactically sugar'ed
+// building of CIRCT RTL dialect operations.
+template <typename CDEValue>
+class Generator {
+public:
+  using CDEPorts = CDEPorts<CDEValue>;
+  Generator(Location loc, OpBuilder &b)
+      : loc(loc), bb(b, loc), ctx(loc, b, bb) {}
+  virtual ~Generator() = default;
+
+protected:
+  /// Unwraps a range of CDEValues to their underlying mlir::Value's.
+  static llvm::SmallVector<Value> unwrap(llvm::ArrayRef<CDEValue> values) {
+    SmallVector<Value> unwrapped;
+    llvm::transform(values, std::back_inserter(unwrapped),
+                    [](auto &v) { return v.get(); });
+    return unwrapped;
+  }
+
+  // Executes an N-ary operation on a range of CDEValues and returns the
+  // result as a CDEValue.
+  template <typename TOp>
+  CDEValue execNAryOp(llvm::ArrayRef<CDEValue> operands) {
+    return CDEValue(
+        &ctx, ctx.b.create<TOp>(operands[0].get().getLoc(), unwrap(operands)));
+  }
+
+  // Various N-ary operations.
+  CDEValue And(llvm::ArrayRef<CDEValue> operands) {
+    return execNAryOp<CDEValue::OpTypes::And>(operands);
+  }
+
+  CDEValue Or(llvm::ArrayRef<CDEValue> operands) {
+    return execNAryOp<CDEValue::OpTypes::Or>(operands);
+  }
+
+  CDEValue Xor(llvm::ArrayRef<CDEValue> operands) {
+    return execNAryOp<CDEValue::OpTypes::XOr>(operands);
+  }
+
+  CDEValue cint(size_t width, int64_t value) {
```

Comment:
nit: cint -> constant and use argument overloading to differentiate between types.